python/python/lance/dataset.py

```diff
@@ -1213,31 +1219,37 @@ def create_scalar_index(
         if column not in self.schema.names:
             raise KeyError(f"{column} not found in schema")
 
-        field = self.schema.field(column)
-        if (
-            not pa.types.is_integer(field.type)
-            and not pa.types.is_floating(field.type)
-            and not pa.types.is_boolean(field.type)
-            and not pa.types.is_string(field.type)
-            and not pa.types.is_temporal(field.type)
-        ):
-            raise TypeError(
-                f"Scalar index column {column} must be int",
-                ", float, bool, str, or temporal",
+        index_type = index_type.upper()
+        if index_type not in ["BTREE", "BITMAP", "TAG"]:
+            raise NotImplementedError(
+                (
+                    'Only "BTREE", "TAG", or "BITMAP" are supported for '
+                    f"scalar columns.  Received {index_type}",
+                )
             )
 
+        field = self.schema.field(column)
+        if index_type in ["BTREE", "BITMAP"]:
+            if (
+                not pa.types.is_integer(field.type)
+                and not pa.types.is_floating(field.type)
+                and not pa.types.is_boolean(field.type)
+                and not pa.types.is_string(field.type)
+                and not pa.types.is_temporal(field.type)
+            ):
+                raise TypeError(
+                    f"BTREE/BITMAP index column {column} must be int",
+                    ", float, bool, str, or temporal",
+                )
+        elif index_type == "TAG":
+            if not pa.types.is_list(field.type):
```

Comment:
I wonder if we want to support `str` columns. this seems very similar to the idea of "Symbol" in questdb: https://questdb.io/docs/concept/symbol/

Reply:
`Symbol` looks like Arrow's `Dictionary` type?

This supports `List<String>`.  I'm not sure `String` (without the `List`) makes sense here.  The functions "array_has_any" and "array_has_all" are not used with a `String` column.  The closest analogue might be storing a bitmap per character or bitmap per word/token but we need a valid use case first (and FTS indices are probably better for that type of query).

Reply:
Also, wrt `Symbol`, another TODO I plan to support this week is support for indices on dictionary columns.  I think the customer in this particular scenario is using `List<Dictionary<String>>` so we eventually want to get there.  `Dictionary<String>` is good for storage purposes but won't have much impact on index performance (unless using btree).